### ISSUE TITLE
Improve `noUncheckedIndexedAccess` with `array.length`

### ISSUE THREAD

**Issue by yw662** [microsoft/TypeScript#53678]
# Suggestion

## 🔍 Search Terms
noUncheckedIndexedAccess, length, for loop

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Make `noUncheckedIndexedAccess` work slightly better with `array.length`

## 📃 Motivating Example

eg. in the code below,
```ts
const array: number[] = [0, 1, 2]
let n = 0
for(let i = 0; i < array.length; i++) {
  n = n + array[i]
}
```

will cause false positive with `noUncheckedIndexedAccess` saying `array[i]` can be `undefined`. However it cannot since we have condition `i < array.length`.

And another example,
```ts
function f(arr: number[]) {
  if(arr.length < 1) {
    ...
  } else {
    const v: number = arr[0]
    ...
  }
}
```
We have `arr.length >= 1` at the very line so `arr[0]` is number for sure. 

Doing math in type checks is generally troublesome but if it is restricted to some particular cases it can be very helpful.

## 💻 Use Cases

`noUncheckedIndexedAccess` are not used very much because it can cause lots of false positives. Fixing at least some of the issues can make it much more useable.

**Comment by MartinJohns** [comment]
Sounds like a duplicate of #38000.

**Comment by yw662** [comment]
> Sounds like a duplicate of #38000.

No this is particularly about `noUncheckedIndexedAccess`.

**Comment by RyanCavanaugh** [comment]
Our goal with `noUncheckedIndexedAccess` was to give you a fully-sound representation of indexing. Given that we can't reliably track external mutation of an array, any check on `length` is only valid for an extremely short window of execution. The trade-off for that is indeed false positives; this is the trade-off that you're taking when you turn that option on.

**Comment by fatcerberus** [comment]
Also note that `arr.length > 0` doesn't guarantee that `arr[0]` is defined (sparse arrays are evil).

**Comment by yw662** [comment]
> is only valid for an extremely short window of execution.

But can it be a `the next very line` ?

**Comment by yw662** [comment]
> Also note that `arr.length > 0` doesn't guarantee that `arr[0]` is defined (sparse arrays are evil).

That is true for `(number | undefined)[]` but not for `number[]`.Arrays typed `T[]` are not sparse.

**Comment by fatcerberus** [comment]
@yw662 

```ts
const x: number[] = [];
x[42] = 0;
console.log(x.length);
console.log(x[0]);
// output: 43, undefined
```

[Playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/MYewdgzgLgBAHgLhmArgWwEYFMBOBtAXRgF4ZCBuAKDjwBYAmI0gBitEhABssA6TkAOYAKOHyxgBUABYBKNuAhde-YTWYE5lIA)
This is not caught by `noUncheckedIndexAccess`.

**Comment by yw662** [comment]
> @yw662 
> 
> ```ts
> const x: number[] = [];
> x[42] = 0;
> console.log(x.length, x[0]);
> // output: 43, undefined
> ```
> 
> [Playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/MYewdgzgLgBAHgLhmArgWwEYFMBOBtAXRgF4ZCBuAKDjwBYAmI0gBitEhABssA6TkAOYAKOHyxgBUABYAaeHmYEAlFSA)
> 
> This is not caught by `noUncheckedIndexedAccess`.

Oh `noUncheckedIndexedAccess` would allow that :-).
I think I will create another issue for or @RyanCavanaugh .

**Comment by WoodyWoodsta** [comment]
How about:

```ts
const shouldBeDefined = arr[arr.length - 1];
```

Is there anything that can get in-between the length access and the array access in this scenario?

**Comment by RyanCavanaugh** [comment]
`arr` could have length 0
